// clarity-angular/datagrid/datagrid-footer.js
/*
 * Copyright (c) 2016-2017 VMware, Inc. All Rights Reserved.
 * This software is released under MIT license.
 * The full license information can be found in LICENSE in the root directory of this project.
 */
import { Component } from "@angular/core";
import { Selection, SelectionType } from "./providers/selection";
var DatagridFooter = (function () {
    function DatagridFooter(selection) {
        this.selection = selection;
        /* reference to the enum so that template can access */
        this.SELECTION_TYPE = SelectionType;
    }
    return DatagridFooter;
}());
export { DatagridFooter };
DatagridFooter.decorators = [
    { type: Component, args: [{
                selector: "clr-dg-footer",
                template: "\n        <ng-container \n            *ngIf=\"(selection.selectionType === SELECTION_TYPE.Multi) && (selection.current.length > 0)\">\n            <clr-checkbox [clrDisabled]=\"true\" [clrChecked]=\"true\" class=\"datagrid-footer-select\">\n                {{selection.current.length}}\n            </clr-checkbox>\n        </ng-container>\n        <ng-content></ng-content>\n    ",
                host: {
                    "[class.datagrid-foot]": "true",
                }
            },] },
];
/** @nocollapse */
DatagridFooter.ctorParameters = function () { return [
    { type: Selection, },
]; };
